applications/StructuralMechanicsApplication/tests/cpp_tests/test_small_strain_isotropic_damage_3d.cpp

```diff
@@ -0,0 +1,274 @@
+
+namespace Kratos
+{
+namespace Testing
+{
+// We test the associated plasticity Constitutive laws...
+typedef Node<3> NodeType;
+
+// Check the correct calculation of the integrated stress with the CL's in small strain
+KRATOS_TEST_CASE_IN_SUITE(_ConstitutiveLaw_SmallStrainIsotropicDamage3D, KratosStructuralMechanicsFastSuite)
+{
+    ConstitutiveLaw::Parameters cl_parameters;
+    Properties material_properties;
+    Vector stress_vector(6), strain_vector(6);
+    Matrix const_matrix;
+    // Create gauss point
+    Model current_model;
+    ModelPart& test_model_part = current_model.CreateModelPart("Main");
+    NodeType::Pointer p_node_1 = test_model_part.CreateNewNode(1, 0.0, 0.0, 0.0);
+    NodeType::Pointer p_node_2 = test_model_part.CreateNewNode(2, 1.0, 0.0, 0.0);
+    NodeType::Pointer p_node_3 = test_model_part.CreateNewNode(3, 0.0, 1.0, 0.0);
+    NodeType::Pointer p_node_4 = test_model_part.CreateNewNode(4, 0.0, 0.0, 1.0);
+    Tetrahedra3D4<NodeType> geometry = Tetrahedra3D4<NodeType>(p_node_1, p_node_2, p_node_3, p_node_4);
+    // Set material properties
+    material_properties.SetValue(YOUNG_MODULUS, 3000);
+    material_properties.SetValue(POISSON_RATIO, 0.3);
+    material_properties.SetValue(YIELD_STRESS, 0.5);
+    material_properties.SetValue(INFINITY_YIELD_STRESS, 0.7);
+    material_properties.SetValue(ISOTROPIC_HARDENING_MODULUS, 0.3);
+    // Set constitutive law flags:
+    Flags& ConstitutiveLawOptions=cl_parameters.GetOptions();
+    ConstitutiveLawOptions.Set(ConstitutiveLaw::USE_ELEMENT_PROVIDED_STRAIN, true);
+    ConstitutiveLawOptions.Set(ConstitutiveLaw::COMPUTE_STRESS, true);
+    ConstitutiveLawOptions.Set(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR, true);
+    // Set required constitutive law parameters:
+    cl_parameters.SetElementGeometry(geometry);
+    cl_parameters.SetProcessInfo(test_model_part.GetProcessInfo());
+    cl_parameters.SetMaterialProperties(material_properties);
+    cl_parameters.SetStrainVector(strain_vector);
+    cl_parameters.SetStressVector(stress_vector);
+    cl_parameters.SetConstitutiveMatrix(const_matrix);
+    // Create the CL
+    SmallStrainIsotropicDamage3D cl = SmallStrainIsotropicDamage3D();
+
+    // Set variables for the test
+    const double tolerance = 1.0e-4;
+    std::size_t nr_ts = 4;  // timesteps tested
+    Vector ser(nr_ts), dvr(nr_ts);  // reference strain energy, damage variable
+    Matrix epr(nr_ts, 6), str(nr_ts, 6);  // reference strain ("epsilon"), stress
+    Matrix cmr(nr_ts, 6 * 6);  // reference constitutive matrix
+
+
+    //
+    // Test: check correct behavior of internal and calculated variables
+    //
+    KRATOS_CHECK(cl.Has(INELASTIC_FLAG));  // = True - TO BE REMOVED
+    KRATOS_CHECK_IS_FALSE(cl.Has(STRAIN_ENERGY));  // = False, in order to use CalculateValue())
+    KRATOS_CHECK_IS_FALSE(cl.Has(DAMAGE_VARIABLE));  // = False, in order to use CalculateValue())
+
+
+    //
+    // Test: load - unload in traction
+    //
```

Comment:
This test looks really fragile

Reply:
Do you have a suggestion for making it more robust? Here I am reproducing the python test (passing a strain and checking the stress received), but much more comprehensibly, as now I check the constitutive tensor received, calculated values, proper behavior of Has and CalculateValue (problems in the past), and the strain is computed at key values now. (Thanks for the prompt review BTW)